### DISCUSSION TITLE
Issue during build step on in the context of a cloud workflow

### DISCUSSION THREAD

**Discussion by zachcp** [observablehq/framework#1096] [Q&A] [answered]
Hi Team Observable,

As a follow up to #1061 , I've tried to build a report as part of a larger pipeline. In this case the workflow looks something like `do-big-job-on-the-cloud-and-generate-some-files. ---> create beautiful, interactive report using Framework.`   We are using NextFlow on AWS with an FsxDrive backed-by and syncing-to AWS-S3.  My test workflow works locally (native) and locally (docker) but fails in the cloud with error below. 

[ Edit: the issue is in the rename call during build. ] ~~The engineers on my team believe the root cause of the error is that the cloud file systems we are using (and possibly S3 itself) do not like one or more characters in the build directory and this shows up as a `symlink` fail.  The character at issue is quite possibly the `@` character in the `@observablehq` package.~~  I will report back if I have more concrete evidence but I wanted to raise this here as this would be a tough issue to work around on our end and this might be encountered by other users.

Thanks again,
zach cp

```sh
  npm ERR! code ENOSYS
  npm ERR! syscall symlink
  npm ERR! path ../@observablehq/framework/dist/bin/observable.js
  npm ERR! dest /fusion/s3/<our-bucket>/_nextflow/workdir/stage-b31b47f5-d8dc-44f9-aef8-83cb47e64871/e0/bf91de07290249b53b6f4206c2a41c/report/node_modules/.bin/observable
  npm ERR! errno -38
  npm ERR! ENOSYS: function not implemented, symlink '../@observablehq/framework/dist/bin/observable.js' -> '/fusion/s3/<our-bucket>/_nextflow/workdir/stage-npm_***/e0/bf91de07290249b53b6f4206c2a41c/report/node_modules/.bin/observable'
  npm ERR! A complete log of this run can be found in: /root/.npm/_logs/2024-03-15T21_01_55_341Z-debug-0.log
```

**Comment by Fil** [comment]
Thanks for the report. Does the complete log show a trace of where this syscall symlink originates?

**Reply by zachcp** [reply]
I'll see if I can track it down.

**Reply by zachcp** [reply]
I will keep trying but the npm log is generated in the docker image and isn't captured. At the moment the errors above are the best I can do.

**Reply by zachcp** [reply]
These are the NPM sys calls: `rename` and `rmdir`


```sh
npm WARN cleanup       errno: -39,
npm WARN cleanup       code: 'ENOTEMPTY',
npm WARN cleanup       syscall: 'rmdir',
```

```sh
npm ERR! code ENOTEMPTY
npm ERR! syscall rename
npm ERR! path /fusion/s3/<my bucket>/_nextflow/workdir/stage-a4dd9ef8-3691-4101-a134-b4899710a4dc/c0/7798d1f1dcfb13148442dc37963585/report/node_modules/.markdown-it-4DjFjFN1/node_modules/argparse/lib
npm ERR! dest /fusion/s3/<my bucket>/_nextflow/workdir/stage-a4dd9ef8-3691-4101-a134-b4899710a4dc/c0/7798d1f1dcfb13148442dc37963585/report/node_modules/markdown-it/node_modules/argparse/lib
npm ERR! errno -39
npm ERR! ENOTEMPTY: directory not empty, rename '/fusion/s3/<my bucket>/_nextflow/workdir/stage-npm_***/c0/7798d1f1dcfb13148442dc37963585/report/node_modules/.markdown-it-4DjFjFN1/node_modules/argparse/lib' -> '/fusion/s3/<<my bucket>>/_nextflow/workdir/stage-npm_***/c0/7798d1f1dcfb13148442dc37963585/report/node_modules/markdown-it/node_modules/argparse/lib'
npm ERR! A complete log of this run can be found in: /root/.npm/_logs/2024-03-18T16_29_17_180Z-debug-0.log
```

**Reply by zachcp** [reply]
Note the errors changed slightly as I turned off symlinking: ```npm_config_bin_links=false```

**Comment by zachcp** [comment, marked as answer]
So I figured out whats going on , more or less: 

- in most workflows the Docker image is completely pre-compiled
- the workdir is therefore just for scratch.
- here there is the whole `npm install` dance
- our setup doens't like: `rifraf` and `symlinks`
- but you can get the code to work as needed by working around them using something like:

```sh
# note that we are not allowing node t set symlinks
# and that this requires us to invoke observable directly rather than `npm run`
 npm install --no-bin-links && \
  ./node_modules/@observablehq/framework/dist/bin/observable.js build
```
